mlir/lib/Dialect/Arith/Transforms/UnsignedWhenEquivalent.cpp

```diff
@@ -123,35 +152,32 @@ struct ArithUnsignedWhenEquivalentPass
     if (failed(solver.initializeAndRun(op)))
       return signalPassFailure();
 
-    ConversionTarget target(*ctx);
-    target.addLegalDialect<ArithDialect>();
-    target.addDynamicallyLegalOp<DivSIOp, CeilDivSIOp, FloorDivSIOp, RemSIOp,
-                                 MinSIOp, MaxSIOp, ExtSIOp>(
-        [&solver](Operation *op) -> std::optional<bool> {
-          return failed(staticallyNonNegative(solver, op));
-        });
-    target.addDynamicallyLegalOp<CmpIOp>(
-        [&solver](CmpIOp op) -> std::optional<bool> {
-          return failed(isCmpIConvertable(solver, op));
-        });
+    DataFlowListener listener(solver);
 
     RewritePatternSet patterns(ctx);
-    patterns.add<ConvertOpToUnsigned<DivSIOp, DivUIOp>,
-                 ConvertOpToUnsigned<CeilDivSIOp, CeilDivUIOp>,
-                 ConvertOpToUnsigned<FloorDivSIOp, DivUIOp>,
-                 ConvertOpToUnsigned<RemSIOp, RemUIOp>,
-                 ConvertOpToUnsigned<MinSIOp, MinUIOp>,
-                 ConvertOpToUnsigned<MaxSIOp, MaxUIOp>,
-                 ConvertOpToUnsigned<ExtSIOp, ExtUIOp>, ConvertCmpIToUnsigned>(
-        ctx);
-
-    if (failed(applyPartialConversion(op, target, std::move(patterns)))) {
+    populateUnsignedWhenEquivalentPatterns(patterns, solver);
+
+    GreedyRewriteConfig config;
```

Comment:
Do we expect to have to iterate to convergence here? Otherwise can we set options to limit to a single iteration?

Reply:
I think, it should finish in single iteration.

Reply:
Actually, adding `maxIterations = 1` causing it to fail to converge. I think I will leave things as is for now.

Reply:
It fails to converge  (i.e. `applyPatternsAndFoldGreedily` returns failure) even with
```
    config.maxIterations = 1;
    config.useTopDownTraversal = true;
```

Reply:
This is mostly test pass anyway so I don't see much problem here. Downstream we plan to combine it with other patterns. And using greedy driver here may not be ideal, but it's still better than current dialect conversion driver.

Reply:
If this is a test pass, it should be moved to the test folder and be named accordingly though.

Reply:
This isn't intended as a test pass - you're meant to be able to run this (barring the usual philosophical disagreements about even having non-test passes upstream)

Reply:
I wrote this code using the dialect converter because I wanted a one-shot "walk this function exactly once and apply matching patterns"

Reply:
> This isn't intended as a test pass 

Well, then my point stands: we shouldn't involve the greedy rewriter here.